libp2p/protocols/connectivity/autorelay.nim

```diff
@@ -0,0 +1,110 @@
+# Nim-LibP2P
+# Copyright (c) 2022 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import chronos, chronicles, times, tables, sequtils, options
+import ../../switch,
+       relay/[client, utils]
+
+logScope:
+  topics = "libp2p autorelay"
+
+type
+  OnReservationHandler = proc (ma: MultiAddress): Future[void] {.gcsafe, raises: [Defect].}
+
+  AutoRelay* = ref object of Service
+    running: bool
+    runner: Future[void]
+    client: RelayClient
+    npeers: int
+    peers: Table[PeerId, Future[void]]
+    peerJoined: AsyncEvent
+    onReservation: OnReservationHandler
+
+proc reserveAndUpdate(self: AutoRelay, relayPid: PeerId, selfPid: PeerId) {.async.} =
+  while self.running:
+    let
+      rsvp = await self.client.reserve(relayPid).wait(chronos.seconds(5))
+      relayedAddr = MultiAddress.init($(rsvp.addrs[0]) &
+                                  "/p2p-circuit/p2p/" &
+                                  $selfPid).tryGet()
+    await self.onReservation(relayedAddr)
+    await sleepAsync chronos.seconds(rsvp.expire.int64 - times.now().utc.toTime.toUnix)
+
+method setup*(self: AutoRelay, switch: Switch) {.async, gcsafe.} =
+  if self.inUse:
+    warn "Autorelay setup has already been called"
+    return
+  self.inUse = true
+  proc handlePeer(peerId: PeerId, event: PeerEvent) {.async.} =
+    if event.kind == Left and peerId in self.peers:
+      self.peers[peerId].cancel()
+    elif event.kind == Joined and self.peers.len < self.npeers:
+      self.peerJoined.fire()
+  switch.addPeerEventHandler(handlePeer, Joined)
+  switch.addPeerEventHandler(handlePeer, Left)
+
+method innerRun(self: AutoRelay, switch: Switch) {.async, gcsafe.} =
+  while true:
+    # Remove peers that failed
+    var peersToRemove: seq[PeerId]
+    for k, v in self.peers:
+      if v.finished():
+        peersToRemove.add(k)
```

Comment:
can we replace this line by `self.peers.del(k)` and get rid of `peersToRemove`?

Reply:
No it doesn't work, it raises `the length of the table changed while iterating over it [AssertionDefect]`.
And as there's no removeIf in std/tables, I had to do something inelegant.